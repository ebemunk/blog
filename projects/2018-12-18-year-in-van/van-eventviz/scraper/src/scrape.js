import axios from 'axios'
import cheerio from 'cheerio'
import { map } from 'bluebird'

import { getPool, insertObj } from './db'
import { prettyJson, logger } from './util'

const baseUrl = 'https://www.straight.com'
const eventsUrl = `${baseUrl}/listings/events`

const log = logger({
  slack: {
    username: 'Scraper Dude',
    icon_emoji: ':japanese_ogre:',
  },
})

export default async function scrape(args, opts) {
  log.verbose({
    type: 'env',
    env: process.env,
  })
  const pool = getPool()
  let page = 0
  let events = []
  let totalEvents = 0
  let totalSaves = 0
  do {
    events = await parsePage(page)
    const saves = await map(events, e => save(pool, e))
    totalSaves += saves.filter(e => e).length
    totalEvents += events.length
    log.verbose({
      type: 'page',
      page,
      events: events.length,
      saveCount: saves.filter(e => e).length,
    })
    page++
    // events = []
  } while (events.length)
  log.info({
    type: 'scrape',
    pages: page,
    totalEvents,
    totalSaves,
  })
}

export async function parsePage(page) {
  const { data } = await axios.get(`${eventsUrl}?page=${page}`)
  const $ = cheerio.load(data)
  return $('article.cf.event')
    .map((i, el) => {
      const $el = $(el)
      const find = e => $el.find(e)
      return {
        date: find('.date-tag')
          .text()
          .trim(),
        image: find('.card--rounded a img').attr('src'),
        title: find('.title a').text(),
        url: baseUrl + find('.title a').attr('href'),
        id: find('.title a')
          .attr('href')
          .split('/')
          .pop(),
        venue: find('.event--venue')
          .text()
          .trim(),
        tags: find('.event--tags a')
          .map((i, tag) => $(tag).text())
          .toArray(),
      }
    })
    .toArray()
}

export async function save(pool, event) {
  try {
    return await pool.query(
      ...insertObj(
        {
          id: event.id,
          evt: event,
        },
        'events',
      ),
    )
  } catch (e) {
    // 23505 = unique_violation
    if (e.code === '23505') {
      log.verbose({
        type: 'unique_violation',
        eventId: event.id,
      })
    } else {
      e.event = event
      log.verbose({
        type: 'db save error',
        e,
      })
    }
    return false
  }
}
