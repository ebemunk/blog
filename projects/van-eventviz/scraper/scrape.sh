#!/bin/bash
cd "$(dirname "$0")"
git pull
yarn
yarn start scrape &>> /tmp/scrape.log
sleep 150
yarn start details &>> /tmp/details.log
