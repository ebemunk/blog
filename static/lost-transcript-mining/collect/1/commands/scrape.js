import fs from 'fs'

import Promise from 'bluebird'
import _ from 'lodash'
import c from 'chalk'
import ProgressBar from 'progress'

import {
	parseEpisodeIndex,
	saveEpisodeHtml,
	getLines,
	parseLine,
	getActsAndFlashbacks
} from '../lib/scrape'

const log = _.partial(console.log, c.bgGreen('scrape'))
const readFile = Promise.promisify(fs.readFile)
const writeFile = Promise.promisify(fs.writeFile)

export default async function scrape(opts) {
	const {
		concurrency,
		useCache
	} = opts

	log('starting', opts)
	log('getting episode index...')

	const episodeIndex = await readFile('tmp/episode-index.html')
	const episodes = await parseEpisodeIndex(episodeIndex)

	log(`found ${c.blue(episodes.length)} episodes`)

	let progress = new ProgressBar(':current/:total :bar :eta', {
		total: episodes.length,
	})

	await Promise.map(episodes, async (episode) => {
		const filename = `tmp/html/${episode.season}-${episode.episode}.html`
		if( useCache ) {
			episode.html = await readFile(filename)
		} else {
			episode.html = await saveEpisodeHtml(episode.url, filename)
		}

		const lines = getActsAndFlashbacks(getLines(episode.html))
		.map((line, i) => ({
			...line,
			season: episode.season,
			episode: episode.episode,
			seq: i+1
		}))

		await writeFile(`tmp/json/${episode.season}-${episode.episode}.json`, JSON.stringify(lines, null, 2))

		progress.tick()
	}, { concurrency })
}
