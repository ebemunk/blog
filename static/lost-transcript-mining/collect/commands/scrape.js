import fs from 'fs'

import Promise from 'bluebird'
import _ from 'lodash'
import c from 'chalk'
import ProgressBar from 'progress'

import {
	parseEpisodeIndex,
	saveEpisodeHtml,
	getLines,
	parseLine,
} from '../lib/scrape'

const log = _.partial(console.log, c.bgGreen('scrape'))
const readFile = Promise.promisify(fs.readFile)
const writeFile = Promise.promisify(fs.writeFile)

export default async function scrape(opts) {
	const {
		concurrency,
		useCache
	} = opts

	log('starting')
	log('getting episode index...')

	const episodeIndex = await readFile('tmp/episode-index.html')
	const episodes = await parseEpisodeIndex(episodeIndex)

	log(`found ${c.blue(episodes.length)} episodes`)

	let progress = new ProgressBar(':current/:total :bar :eta', {
		total: episodes.length,
	})

	await Promise.map(episodes, async (episode) => {
		if( useCache ) {
			episode.html = await readFile(`tmp/html/${episode.season}-${episode.episode}.html`)
		} else {
			episode.html = await saveEpisodeHtml(episode.url, `tmp/html/${episode.season}-${episode.episode}.html`)
		}

		const lines = getLines(episode.html)
		.map((line, i) => ({
			...line,
			season: episode.season,
			episode: episode.episode,
			seq: i+1
		}))

		await writeFile(`tmp/json/${episode.season}-${episode.episode}.json`, JSON.stringify(lines, null, 2))

		progress.tick()
	}, { concurrency })
}
