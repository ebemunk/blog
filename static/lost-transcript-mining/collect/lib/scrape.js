import fs from 'fs'

import Promise from 'bluebird'
import axios from 'axios'
import cheerio from 'cheerio'
import _ from 'lodash'

const writeFile = Promise.promisify(fs.writeFile)

export async function parseEpisodeIndex(html) {
	const $ = cheerio.load(html)
	const seasons = _.range(1,7)
	.map(season => {
		const episodes = $(`#Season_${season}`)
		// parsing the cheerio tree
		.parent()
		.next('ol')
		.children()
		.get()
		.map((el, i) => {
			const link = $('a', el)
			// an array of these are returned
			return {
				season,
				episode: i+1,
				name: link.attr('title').replace(' transcript', ''),
				url: link.attr('href')
			}
		})
		return episodes
	})
	// flatten
	return seasons.reduce((prev, cur) => prev.concat(cur))
}

export async function saveEpisodeHtml(url, path) {
	const { data } = await axios.get(`${url}`)
	await Promise.fromCallback(cb => fs.writeFile(path, data, cb))
	return data
}

export function getLines(htmlString) {
	const $ = cheerio.load(htmlString)
	const lines = $('#toc')
	// everything after table of contents
	.nextAll()
	// only hr h2 and p tags with content in them
	.filter((i, el) => {
		return ['hr', 'h2', 'p'].includes(el.tagName) && $(el).text().trim()
	})
	.map((i, el) => {
		return {
			tagName: el.tagName,
			text: $(el).text().trim().replace(/\n/g, '')
		}
	})
	.get()
	.map(lineType)
	.filter(Boolean)
	return lines
}

export function lineType({tagName, text}) {
	switch( tagName ) {
		case 'hr': return {
			type: 'scene'
		}

		case 'h2': return {
			type: 'act'
		}

		case 'p': {
			const stageDirection = text.match(/^\[([^\]]+)\]?(?: (.+))?$/)

			if( stageDirection ) {
				return {
					type: 'stageDirection',
					raw: text,
					line: stageDirection[1]
				}
			} else {
				const parsed = parseLine(text)
				if( ! parsed ) return false
				return {
					type: 'dialog',
					raw: text,
					...parsed
				}
			}
		}
	}
}

export function parseLine(text) {
	let line = text
	// everything inside square brackets or paranthesis is a direction
	let directions = []
	// [ and ]
	.concat(text.match(/\[([^\]]+)\]\.?/g))
	// ( and )
	.concat(text.match(/\(([^\]]+)\)\.?/g))
	// remove falsy
	.filter(Boolean)
	// strip and remove from source text
	.map(str => {
		line = line.replace(str, '').replace('  ', ' ')
		return str.replace(/[\[\]\(\)]/g, '').replace(/\.$/, '')
	})

	const charLine = line.trim().match(/^(.+?)\:\s*(.+)?$/)
	if( ! charLine ) {
		console.log('unparseable line:', text)
		console.log()
		return
	}
	const char_name = charLine[1].trim().toUpperCase()
	line = charLine[2] || ''
	return {
		char_name,
		line,
		directions,
	}
}
