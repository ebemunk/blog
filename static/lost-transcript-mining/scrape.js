import fs from 'fs'

import _ from 'lodash'
import Promise from 'bluebird'
import ProgressBar from 'progress'
import c from 'chalk'

import { getSeasonList, getEpisode } from './scrape-fns'

const log = _.partial(console.log, c.bgGreen('scraper'))
const writeFile = Promise.promisify(fs.writeFile)
const pad2 = _.partialRight(_.padStart, 2, 0)

async function scrape() {
	log('starting')
	const seasons = await getSeasonList()
	const allEpisodes = seasons.reduce((prev, cur) => {
		return [...prev, ...cur.episodes]
	}, [])
	log(`found ${c.blue(allEpisodes.length)} episodes in ${c.blue(seasons.length)} seasons`)
	log('downloading...')
	// const progress = new ProgressBar(':current/:total :bar :eta', allEpisodes.length)
  // const kek = _.find(allEpisodes, {season: 6, ep: 5})
  return Promise.map(allEpisodes, async episode => {
    log(episode)
		const episodeText = await getEpisode(episode)
		const fileName = `episodes/${pad2(episode.season)}-${pad2(episode.ep)}.json`
		await writeFile(fileName, JSON.stringify(episodeText, null, 2))
		// progress.tick()
	}, {concurrency: 10})
}

scrape()
.then(() => {
	log('done');
})
.catch(err => {
	log(err);
})
