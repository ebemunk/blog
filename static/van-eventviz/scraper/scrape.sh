git pull
yarn
yarn start scrape
