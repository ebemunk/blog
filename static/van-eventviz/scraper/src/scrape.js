import axios from 'axios'
import cheerio from 'cheerio'
import * as R from 'ramda'

import { logger } from './util'

const log = logger('scrape')

const baseUrl = 'https://www.straight.com'
const eventsUrl = `${baseUrl}/listings/events`

export default async function scrape(args, opts) {
  const { data } = await axios.get(eventsUrl)
  return listPage(data)
}

export function listPage(html) {
  const $ = cheerio.load(html)
  const events = $('.teaser--summary')

  return events
    .map((i, el) => {
      const $el = $(el)
      const find = e => $el.find(e)
      return {
        date: find('.date-plate').text(),
        image: find('figure a img').attr('src'),
        title: find('.title a').text(),
        url: baseUrl + find('.title a').attr('href'),
        venue: find('.event--venue').text().replace('Where: ', ''),
        tags: find('.venue--tags a').map((i, tag) => $(tag).text()).toArray()
      }
    })
    .toArray()
}
