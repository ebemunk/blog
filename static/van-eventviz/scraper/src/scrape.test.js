import { readFile } from 'fs'
import { fromCallback } from 'bluebird'
import axios from 'axios'

import scrape from './scrape'

describe('scrape', () => {
  let data

  beforeAll(async () => {
    data = await fromCallback(cb => readFile('./test/test2.html', 'utf-8', cb))
  })

  it('gets', async () => {
    axios.get = jest.fn(() => Promise.resolve({data}))
    const events = await scrape()
    expect(events).toMatchSnapshot()
  })
})
