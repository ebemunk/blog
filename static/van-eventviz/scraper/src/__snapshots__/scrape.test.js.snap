// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scrape gets 1`] = `
Array [
  Object {
    "date": "Jan 13 Sat",
    "image": "//d2ciprw05cjhos.cloudfront.net/files/v3/styles/gs_small/public/events/17/08/yoga2_170816.jpg?itok=1dLi_cwS",
    "tags": Array [],
    "title": "240hr Enriched Yoga Teacher Training",
    "url": "https://www.straight.com/listings/events/835816",
    "venue": "Vancouver School of Yoga",
  },
  Object {
    "date": "Jan 13 Sat",
    "image": "//d2ciprw05cjhos.cloudfront.net/files/v3/styles/gs_small/public/events/17/11/above_the_hospital_publicity_image9.jpg?itok=5-UCSzIv",
    "tags": Array [
      "Theatre",
      "Performing Arts",
    ],
    "title": "Above the Hospital",
    "url": "https://www.straight.com/listings/events/1000901",
    "venue": "The Red Gate Revue Stage",
  },
  Object {
    "date": "Jan 13 Sat",
    "image": "//d2ciprw05cjhos.cloudfront.net/files/v3/styles/gs_small/public/events/18/01/image0011.png?itok=QmPbaPVx",
    "tags": Array [
      "Film",
      "Kids' Stuff",
    ],
    "title": "Amazon Adventure in IMAX film",
    "url": "https://www.straight.com/listings/events/1015826",
    "venue": "Science World at Telus World of Science",
  },
  Object {
    "date": "Jan 13 Sat",
    "image": "//d2ciprw05cjhos.cloudfront.net/files/v3/styles/gs_small/public/events/17/02/bird_necklace_rs.jpg?itok=Z9fUxWrA",
    "tags": Array [
      "Galleries",
    ],
    "title": "Amazonia: The Rights of Nature",
    "url": "https://www.straight.com/listings/events/869896",
    "venue": "The Museum of Anthropology at UBC",
  },
  Object {
    "date": "Jan 13 Sat",
    "image": "//d2ciprw05cjhos.cloudfront.net/files/v3/styles/gs_small/public/events/18/01/gregsestero_180104.jpg?itok=KEsx1Fo0",
    "tags": Array [
      "Film",
    ],
    "title": "An Evening With Greg Sestero",
    "url": "https://www.straight.com/listings/events/1014161",
    "venue": "Rio Theatre",
  },
  Object {
    "date": "Jan 13 Sat",
    "image": undefined,
    "tags": Array [
      "Galleries",
      "Other",
    ],
    "title": "Art Rental Show",
    "url": "https://www.straight.com/listings/events/1010851",
    "venue": "Cityscape Community Art Space",
  },
  Object {
    "date": "Jan 13 Sat",
    "image": undefined,
    "tags": Array [
      "Galleries",
      "Attractions",
    ],
    "title": "Barb Pearson and Jodi Stark Exhibit",
    "url": "https://www.straight.com/listings/events/1008761",
    "venue": "District Foyer Gallery",
  },
  Object {
    "date": "Jan 13 Sat",
    "image": "//d2ciprw05cjhos.cloudfront.net/files/v3/styles/gs_small/public/events/17/12/mehra-04_view1.jpeg?itok=bYrRxcmF",
    "tags": Array [
      "Galleries",
      "Activism",
    ],
    "title": "Beginning with the Seventies: GLUT exhibition",
    "url": "https://www.straight.com/listings/events/1009296",
    "venue": "Morris and Helen Belkin Art Gallery",
  },
  Object {
    "date": "Jan 13 Sat",
    "image": "//d2ciprw05cjhos.cloudfront.net/files/v3/styles/gs_small/public/events/17/12/belledejour_171218.jpg?itok=5zw3CHGV",
    "tags": Array [
      "Film",
    ],
    "title": "Belle de Jour",
    "url": "https://www.straight.com/listings/events/1009176",
    "venue": "Vancity Theatre",
  },
  Object {
    "date": "Jan 13 Sat",
    "image": "//d2ciprw05cjhos.cloudfront.net/files/v3/styles/gs_small/public/events/17/12/maki_pink_veil_pixie.jpg?itok=SZK3j60Y",
    "tags": Array [
      "Dance",
      "Performing Arts",
    ],
    "title": "Bellydance Classes with Maki",
    "url": "https://www.straight.com/listings/events/1010471",
    "venue": "Scotiabank Dance Centre",
  },
  Object {
    "date": "Jan 13 Sat",
    "image": undefined,
    "tags": Array [
      "Concerts",
    ],
    "title": "Best in Vancouver Finals",
    "url": "https://www.straight.com/listings/events/1005696",
    "venue": "Railway Stage and Beer Café",
  },
  Object {
    "date": "Jan 13 Sat",
    "image": undefined,
    "tags": Array [
      "Galleries",
      "Attractions",
    ],
    "title": "Bev Ellis | Broken: Together",
    "url": "https://www.straight.com/listings/events/1008771",
    "venue": "City Atrium Gallery",
  },
  Object {
    "date": "Jan 13 Sat",
    "image": "//d2ciprw05cjhos.cloudfront.net/files/v3/styles/gs_small/public/events/18/01/bigjayoakerson_180104.jpg?itok=nj3KGPfg",
    "tags": Array [
      "Comedy",
    ],
    "title": "Big Jay Oakerson",
    "url": "https://www.straight.com/listings/events/1014121",
    "venue": "The Comedy MIX",
  },
  Object {
    "date": "Jan 13 Sat",
    "image": "//d2ciprw05cjhos.cloudfront.net/files/v3/styles/gs_small/public/events/17/11/23843397_1732906543386291_1390639040143937972_n.jpg?itok=O0A3o_Pw",
    "tags": Array [
      "Performing Arts",
      "Food & Drink",
    ],
    "title": "BigTop Tuesdays",
    "url": "https://www.straight.com/listings/events/1000976",
    "venue": "Libra Room",
  },
  Object {
    "date": "Jan 13 Sat",
    "image": "//d2ciprw05cjhos.cloudfront.net/files/v3/styles/gs_small/public/events/18/01/birdboy_180105.jpg?itok=goMh20xc",
    "tags": Array [
      "Film",
    ],
    "title": "Birdboy: The Forgotten Children",
    "url": "https://www.straight.com/listings/events/1014506",
    "venue": "Vancity Theatre",
  },
  Object {
    "date": "Jan 13 Sat",
    "image": "//d2ciprw05cjhos.cloudfront.net/files/v3/styles/gs_small/public/events/18/01/blue_180105.jpg?itok=Vgrbh0Fx",
    "tags": Array [
      "Film",
    ],
    "title": "Blue",
    "url": "https://www.straight.com/listings/events/1014501",
    "venue": "Vancity Theatre",
  },
  Object {
    "date": "Jan 13 Sat",
    "image": undefined,
    "tags": Array [],
    "title": "Blues Brunch With Rob Montgomery",
    "url": "https://www.straight.com/listings/events/1017201",
    "venue": "Railway Stage and Beer Café",
  },
  Object {
    "date": "Jan 13 Sat",
    "image": "//d2ciprw05cjhos.cloudfront.net/files/v3/styles/gs_small/public/events/18/01/bombshell_180105.jpg?itok=EZuWk6JG",
    "tags": Array [
      "Film",
    ],
    "title": "Bombshell: The Hedy Lamarr Story",
    "url": "https://www.straight.com/listings/events/1014496",
    "venue": "Vancity Theatre",
  },
  Object {
    "date": "Jan 13 Sat",
    "image": "//d2ciprw05cjhos.cloudfront.net/files/v3/styles/gs_small/public/events/18/01/books.jpg?itok=RJnwfNl4",
    "tags": Array [
      "Performing Arts",
      "Galleries",
    ],
    "title": "Books: Patrick Blenkarn’s MFA Graduating Project",
    "url": "https://www.straight.com/listings/events/1013541",
    "venue": "Studio D, SFU's Goldcorp Centre for the Arts",
  },
  Object {
    "date": "Jan 13 Sat",
    "image": "//d2ciprw05cjhos.cloudfront.net/files/v3/styles/gs_small/public/events/18/01/top10_180108.jpg?itok=mk73SmwJ",
    "tags": Array [
      "Film",
    ],
    "title": "Canada's Top 10 Film Festival",
    "url": "https://www.straight.com/listings/events/1015526",
    "venue": "The Cinematheque",
  },
  Object {
    "date": "Jan 13 Sat",
    "image": "//d2ciprw05cjhos.cloudfront.net/files/v3/styles/gs_small/public/events/17/11/canyonlights_171120.jpg?itok=4qbCkDy6",
    "tags": Array [
      "Kids' Stuff",
      "Holiday",
    ],
    "title": "Canyon Lights",
    "url": "https://www.straight.com/listings/events/997566",
    "venue": "Capilano Suspension Bridge",
  },
  Object {
    "date": "Jan 13 Sat",
    "image": "//d2ciprw05cjhos.cloudfront.net/files/v3/styles/gs_small/public/events/17/11/canyon_lights_2013-12_preview.jpeg?itok=HkNV7_oO",
    "tags": Array [
      "Other",
    ],
    "title": "Canyon Lights: Vancouver City Capilano Suspension Bridge Park Tour",
    "url": "https://www.straight.com/listings/events/999956",
    "venue": "Canada Place",
  },
  Object {
    "date": "Jan 13 Sat",
    "image": "//d2ciprw05cjhos.cloudfront.net/files/v3/styles/gs_small/public/events/17/10/carolsawyer_171024.jpg?itok=tHnvUvbP",
    "tags": Array [
      "Galleries",
    ],
    "title": "Carol Sawyer: The Natalie Brettschneider Archive",
    "url": "https://www.straight.com/listings/events/985691",
    "venue": "Vancouver Art Gallery",
  },
  Object {
    "date": "Jan 13 Sat",
    "image": undefined,
    "tags": Array [
      "Concerts",
    ],
    "title": "Celebrated pianist Monty Alexander",
    "url": "https://www.straight.com/listings/events/1004156",
    "venue": "Frankie’s Jazz Club",
  },
  Object {
    "date": "Jan 13 Sat",
    "image": undefined,
    "tags": Array [
      "Galleries",
    ],
    "title": "Celebration: Youth Art Exhibition",
    "url": "https://www.straight.com/listings/events/1008276",
    "venue": "Lynn Valley Library Square",
  },
]
`;
